rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+  - As described under [Binders](#binders) below, nested attribute sets are always expanded.
+
+**Examples**
+
+```nix
+#1
+[
+  { }
+  { foo = "bar"; }
+  {
+    foo = {
+      bar = "baz";
+    };
+  }
+  { foo.bar = "baz"; }
+]
+
+#2
+[
+  [ 1 ]
+  [
+    2
+    3
+  ]
+]
+
+#3
+[
+  [
+    1
+    2
+    3
+  ]
+]
+
+#4
+[
+  {
+    mySingletons = [
+      [
+        ({
+          # stuff in there
+        })
+      ]
+    ];
+  
+    mySingletons' = [
+      [
+        (function call)
+      ]
+    ];
+  }
+]
+```
+
+**Drawbacks**
+
+- Singleton lists may use a lot of indentation
+
+**Rationale and alternatives**
+
+- Have a special compact form for singleton lists, to reduce the indentation level and remove two additional lines
+  ```nix
+  foo = [ {
+    # content
+  } ];
+  ```
+- Be a bit less eager about expanding lists and attribute sets (for example by allowing up to n elements on a single line).
+  ```nix
+  [ x y ]
+  ```
+
+### Binders and inherit
+
+Let bindings and attribute sets share the same syntax for their items, which is discussed here together.
+
+#### Binders
+
+**Description**
+
+Binders have the most special cases to accommodate for many common Nixpkgs idioms.
+Generally, the following styles exist, which are used depending on the kind and size of the value:
+
+```nix
+#1 single line
+foo = "bar";
+
+#2 single line, on a new line
+very.long.foo =
+  function arg1 arg2 arg3;
+
+#3 multi line, starting on the same line
+foo = function {
+  # args
+};
+
+#4 multi line, starting on a new line
+foo =
+  function
+    arg1
+    arg2
+    arg3
+;
+
+#5 simple function arguments (<=2) start on the same line
+add = x: y: {
+  result = x + y;
+};
+
+#6 attribute bindings always start on a new line
+# If we add more attributes, the indentation stays the same, see #7
+split =
+  { x, ... }:
+  {
+    inc = x + 1;
+    dec = x - 1;
+  };
+
+#7 with enough attribute bindings, they will start on a new line and become multline 
+# it is common to expand input arguments in a way that does not change the number of function applications necessary
+outputs =
+  {
+    x,
+    y,
+    z,
+    ...
+  }:
+  {
+    result = x + y;
+  };
+```
+
+Notable special cases are:
+
+- Single line values that would not benefit from style #2 keep using #1, even if this makes it go above the line limit. This mostly applies to simple strings and paths.
+- Attribute set values are *always* expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
+  ```nix
+  {
+    foo.bar.baz = "qux";
+    foo' = {
+      bar.baz = "qux";
+    };
+  }
+  ```
+- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
+- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
+  ```nix
+  buildInputs = with pkgs; [
+    some
+    dependencies
+  ];
```

Comment:
This doesn't have to be an exception if there's a general rule for `with` as suggested in https://github.com/NixOS/rfcs/pull/166/files#r1394820149.